snakemake/executors/flux.py

```diff
@@ -0,0 +1,266 @@
+        )
+        self.jobs_formatter = flux.job.JobInfoFormat(jobs_format)
+
+        # Note there is no attr for "id", its always returned
+        fields2attrs = {
+            "id.f58": (),
+            "username": ("userid",),
+            "exception.occurred": ("exception_occurred",),
+            "exception.type": ("exception_type",),
+            "exception.note": ("exception_note",),
+            "runtime": ("t_run", "t_cleanup"),
+            "status": ("state", "result"),
+            "status_abbrev": ("state", "result"),
+            "t_remaining": ("expiration", "state", "result"),
+        }
+
+        # Set job attributes we will use later to get job statuses
+        self.job_attrs = set()
+        for field in self.jobs_formatter.fields:
+            if field not in fields2attrs:
+                self.job_attrs.update((field,))
+            else:
+                self.job_attrs.update(fields2attrs[field])
+
+    def _init_flux(self):
+        """
+        Ensure we can import flux in the environment.
+        """
+        if not flux:
+            sys.exit(
+                "Cannot import flux. Is a cluster available to you with Python bindings?"
+            )
+        self._fexecutor = flux.job.FluxExecutor()
+
+    def cancel(self):
+        """
+        cancel execution, usually by way of control+c. Cleanup is done in
+        shutdown (deleting cached workdirs in Google Cloud Storage
+        """
+        for job in self.active_jobs:
+            if not job.flux_future.done():
+                flux.job.cancel(self.f, job.jobid)
+        self.shutdown()
+
+    def _set_job_resources(self, job):
+        """
+        Given a particular job, generate the resources that it needs,
+        including default regions and the virtual machine configuration
+        """
+        mem_mb = job.resources.get("mem_mb", 15360)
+
+        # Update default resources
+        self.default_resources = DefaultResources(
+            from_other=self.workflow.default_resources
+        )
+        self.default_resources.set_resource("mem_mb", mem_mb)
+
+    def get_snakefile(self):
+        assert os.path.exists(self.workflow.main_snakefile)
+        return self.workflow.main_snakefile.replace(self.workdir, "").strip(os.sep)
+
+    def _get_jobname(self, job):
+        # Use a dummy job name (human readable and also namespaced)
+        return "snakejob-%s-%s-%s" % (self.run_namespace, job.name, job.jobid)
+
+    def _generate_environment(self):
+        """
+        Loop through envvars (keys to host environment) and add
+        any that are requested for the runtime environment
+        """
+        envvars = dict(os.environ)
+        for key in self.envvars:
+            try:
+                envvars[key] = os.environ[key]
+            except Exception:
+                continue
+        return envvars
+
+    def format_job_exec(self, job):
+        """
+        We don't want a python -m to run.
```

Comment:
Why not? Is there any downside? If this is really needed to be different, rather make this specific part a separate overwritable method in the AbstractExecutor, instead of copying the entire method from there.

Reply:
I'm not sure, but in my example the python -m didn't work, but running snakemake did. I think the issue I was running into is having a different flux install (provided in the container) with the one that works with snakemake (via miniconda). I couldn't get them working from the same one - so python -m won't work (the python does not know about snakemake as it's not installed there) but calling the executable directly does. This is likely an artifact of the container I'm using (already built by the flux crew) but if I added python -m it won't work.

Reply:
Mhm, I think we should try to understand that. So, how is that with flux. Do jobs always run inside of a container, or only in case of cloud mode?

Reply:
Flux isn't typically in a container I don't think - this is just the dummy example that I'm testing/developing with! Ping @trws and @grondo I think he can add more insight (I haven't yet used flux outside of this single container demo environment).

Reply:
Flux is a resource manager similar to Slurm, PBS, Torque, etc in that it is designed to run across a set of resources, as on a cluster for example. However, it can also be started as a job within one of those resource managers (including itself), to give full RM capabilities to the job. It can also be started standalone on a node, VM, or container to help manage those resources.

I guess to answer more specifically, Flux is not typically used inside a container, but nothing prevents the user from doing that.

Reply:
@grondo another question - how does flux handle environment? E.g., should we be grabbing the user environment to provide to snakemake, or start with an empty set and add only those envar names requested? I did the first because I guessed that we would want to preserve most about the user path /home, etc., but maybe that's the HPC use case and not for Flux?

Reply:
Flux is more like an HPC resource manager/scheduler,  and most of the command line utilities copy the current environment into the job environment. I'm not familiar with snakemake, but I would assume that would be preferred here as well?

Reply:
By default flux propagates the entire environment captured where `flux run` or similar was invoked, because that tends to be what a user wants, but it provides a variety of ways to cur that down or override it as well.  It's situational enough that I would think we'd probably want to do what a snakemake user would expect, whatever that is.

To answer the earlier question, flux doesn't run jobs in containers, though that could be supported.  It sounds like there were two python environments installed, and the one with flux bindings is different from the one with snakemake installed.  We could look at making it easier to install the flux binding components into the same environment, or snakemake into the one flux is using to get around the `python -m` problem.

Reply:
@trws that would be fantastic - is this something I could help with? I don't remember the exact issues, but I couldn't get snakemake installed into the default Python in the container.

Reply:
Actually, you probably could, we should chat about that separately.  It would take some re-packaging of the flux python binding component so it can build separately from the autotools-based build of flux to make it convenient. That's something we have several customers asking for, so if you're up for it that would be really easy to support.

Reply:
Yep - would love to help with that! Will ping you tomorrow (and leave me a message in chat if you want to show me stuff sooner).

Reply:
Was there any progress with this? I still would like to get rid of this repetition of superclass code.